src/liballoc/allocator.rs

```diff
@@ -0,0 +1,698 @@
+    /// include successor blocks, or virtual memory tricks.)
+    ///
+    /// If this returns `Err`, then the allocator has made no assertion
+    /// about whether the memory block referenced by `ptr` can or cannot
+    /// fit `new_layout`.
+    ///
+    /// In either case, ownership of the memory block referenced by `ptr`
+    /// has not been transferred, and the contents of the memory block
+    /// are unaltered.
+    unsafe fn grow_in_place(&mut self,
+                            ptr: Address,
+                            layout: Layout,
+                            new_layout: Layout) -> Result<(), CannotReallocInPlace> {
+        let _ = ptr; // this default implementation doesn't care about the actual address.
+        debug_assert!(new_layout.size >= layout.size);
+        if new_layout.align != layout.align {
+            return Err(CannotReallocInPlace);
+        }
+        let (_l, u) = self.usable_size(&layout);
+        // _l <= layout.size()                       [guaranteed by usable_size()]
+        //       layout.size() <= new_layout.size()  [required by this method]
+        if new_layout.size <= u {
+            return Ok(());
+        } else {
+            return Err(CannotReallocInPlace);
+        }
+    }
+
+    /// Attempts to shrink the allocation referenced by `ptr` to fit `new_layout`.
+    ///
+    /// * `ptr` must have previously been provided via this allocator.
+    ///
+    /// * `layout` must *fit* the `ptr` (see above). (The `new_layout`
+    ///   argument need not fit it.)
+    ///
+    /// * `new_layout.size()` must not be greater than `layout.size()`
+    ///   (and must be greater than zero).
+    ///
+    /// Behavior undefined if any of the above three constraints are unmet.
+    ///
+    /// If this returns `Ok`, then the allocator has asserted that the
+    /// memory block referenced by `ptr` now fits `new_layout`, and
+    /// thus can only be used to carry data of that smaller
+    /// layout. (The allocator is allowed to take advantage of this,
+    /// carving off portions of the block for reuse elsewhere.) The
+    /// truncated contents of the block within the smaller layout are
+    /// unaltered, and ownership of block has not been transferred.
+    ///
+    /// If this returns `Err`, then the memory block is considered to
+    /// still represent the original (larger) `layout`. None of the
+    /// block has been carved off for reuse elsewhere, ownership of
+    /// the memory block has not been transferred, and the contents of
+    /// the memory block are unaltered.
+    unsafe fn shrink_in_place(&mut self,
+                              ptr: Address,
+                              layout: Layout,
+                              new_layout: Layout) -> Result<(), CannotReallocInPlace> {
+        let _ = ptr; // this default implementation doesn't care about the actual address.
+        debug_assert!(new_layout.size <= layout.size);
+        if new_layout.align != layout.align {
```

Comment:
Can we just require that `new_layout.align == layout.align` for defined behavior? I can't see that this would be hard for users to follow, and it makes sense to me (obviously you won't change the alignment if you don't change the pointer).

This would require adding the check in the default implementation of `realloc`, but that doesn't seem to be a problem.

We could just only require that `new_layout.align <= layout.align`, but I don't see that being useful for users, and it might be annoying for allocator implementations, since different alignments might result in different size classes.